README.md

```diff
@@ -1,18 +1,28 @@
+# Foxy Improvement Backport Branch
+
+This branch contains a special-purpose version of rosbag2 that is intended to be used in a Foxy application.
+
+It contains performance improvements and new features from Galactic development up to March 23, 2021 - with minor tweaks to maintain build and test compatibility against the stable Foxy distribution.
+
+The popularly-requested improvements in this branch were not possible to backport to Foxy in an API/ABI-maintaining way. Instead, this "Foxy Future" branch is provided for users to build from source and get the performance benefits from nearly a year of development after Foxy.
+
+For more context on this conversation, see https://discourse.ros.org/t/fast-forward-merging-rosbag2-master-api-to-foxy/18927
+
+## Usage instructions
+
+* Check out this branch `foxy-future` from the repository `github.com/ros2/rosbag2` into your ROS workspace (or an overlay workspace)
```

Comment:
have we tested the overlay functionality though? I am currently on an OSX machine, so can't easily overlay through the binaries. Is that something we can test through GH actions?

Reply:
Oof - I thought it was working - but `rosbag2_py` won't build because it's picking up the headers from the `/opt/ros/foxy` install before the local workspace headers. I'm not sure exactly how to solve this - it's an ugly problem, and suggests that the other packages may also be picking up the installed binary headers instead of local

Reply:
We could tell users to uninstall rosbag2 packages - which would uninstall `ros-base` / `desktop` - but won't uninstall their installed packages, just put them in the list for `"The following packages were automatically installed and are no longer required"`

That's not an ideal solution though

Reply:
There's a top-level ROS2 issue already open for this: https://github.com/ros2/ros2/issues/1150

I think we could specifically uninstall a list of rosbag2 packages via `dpkg -r ...`. That won't uninstall the rest of ros2

Reply:
Seems like it is due to #698 
I could try to fix it by this saturday or sunday if this PR waits for it

Reply:
`dpkg -r ...` is not a good solution. It needs to be tested but when any update of package which depends on removed one will be released it might lead to apt being broken or removed dependency being installed again

Reply:
Locally I did `apt-get remove "*rosbag2*"` - which worked fine and let the build succeed, but didn't uninstall any of the other packages. The packages are then susceptible to `apt autoremove`, though, at that point.

@hexonxons do you think that #698 can fix it locally? Or would we need to fix Foxy, get that out in a patch release/sync, then this would work better?

Reply:
@emersonknapp 
> Locally I did apt-get remove "*rosbag2*" - which worked fine and let the build succeed, but didn't uninstall any of the other packages. The packages are then susceptible to apt autoremove, though, at that point.

It works okay, I'm talking about what will happen when something will be updated.
E.g during next foxy release some package which depends on any "*rosbag*" package will be updated. Next at some point u will hit `apt update && apt upgrade` and I am not sure how apt will handle it.

> @hexonxons do you think that #698 can fix it locally? Or would we need to fix Foxy, get that out in a patch release/sync, then this would work better?

I've checked https://github.com/ros2/ros2/issues/1150 and it looks like `ament_target_dependencies` is messing with transitive dependencies.
Rosbag2 needs to be inspected if mentioned problem is due to transitive dependencies or due to missing dependencies.

E.g if we have packages a, b, c:
a depends on b
a depends on c
b depends on c

But `a depends on c` is missing from `package.xml` and `CMakeLists.txt` we also might have same problem, cuz when building `a` on CI we still have `c` installed as a `b` dependency.
I am not sure if ament raise warning in that case, I might be wrong.

Anyway, regarding transitive dependency problem, it it right way to wait for upstream to fix it, but it also could be fixed in a not-so-right way by adding internal transitive dependencies to broken packages.

Reply:
OK - if you can make this branch work against binary-installed Foxy underlay with some trick, I am fine with that as long as we are confident it will keep working. I am happy to wait a couple of days for that if you're volunteering to help :)

Reply:
@emersonknapp I've built it inside ros:foxy image: https://github.com/ros2/rosbag2/pull/780
Inside `compile_commands.json` everything seems ok